File: java/query/TypeQLDelete.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.vaticle.typeql.lang.query;

import com.vaticle.typeql.lang.common.exception.TypeQLException;
import com.vaticle.typeql.lang.pattern.statement.ThingStatement;

import javax.annotation.Nullable;
import java.util.List;

import static com.vaticle.typedb.common.collection.Collections.list;
import static com.vaticle.typeql.lang.common.TypeQLToken.Clause.DELETE;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.VARIABLE_OUT_OF_SCOPE_DELETE;
import static java.util.Objects.requireNonNull;

public class TypeQLDelete extends TypeQLWritable.InsertOrDelete {

    TypeQLDelete(MatchClause match, List<ThingStatement<?>> statements) {
        this(match, statements, Modifiers.EMPTY);
    }

    TypeQLDelete(MatchClause match, List<ThingStatement<?>> statements, @Nullable Modifiers modifiers) {
        super(DELETE, requireNonNull(match), validDeleteStatements(match, statements), modifiers == null ? Modifiers.EMPTY : modifiers);
    }

    static List<ThingStatement<?>> validDeleteStatements(MatchClause match, List<ThingStatement<?>> statements) {
        statements.forEach(stmt -> {
            stmt.variables().forEach(var -> {
                if (var.isNamed() && !match.namedVariables().contains(var)) {
                    throw TypeQLException.of(VARIABLE_OUT_OF_SCOPE_DELETE.message(var.reference()));
                }
            });
        });
        return statements;
    }

    public static class Unmodified extends TypeQLDelete implements TypeQLQuery.Unmodified<TypeQLDelete, TypeQLDelete.Sorted, TypeQLDelete.Offset, TypeQLDelete.Limited> {

        public Unmodified(MatchClause match, List<ThingStatement<?>> statements) {
            super(match, statements, Modifiers.EMPTY);
        }

        public TypeQLUpdate.Unmodified insert(ThingStatement<?>... statements) {
            return insert(list(statements));
        }

        public TypeQLUpdate.Unmodified insert(List<ThingStatement<?>> statements) {
            return new TypeQLUpdate.Unmodified(this.match().get(), this.statements, statements);
        }

        @Override
        public TypeQLDelete modifiers(Modifiers modifier) {
            if (modifier.sorting != null) TypeQLQuery.validateSorting(match, modifier.sorting);
            return new TypeQLDelete(match, statements, modifier);
        }

        @Override
        public TypeQLDelete.Sorted sort(Modifiers.Sorting sorting) {
            return new TypeQLDelete.Sorted(this, sorting);
        }

        @Override
        public TypeQLDelete.Offset offset(long offset) {
            return new TypeQLDelete.Offset(this, offset);
        }

        @Override
        public TypeQLDelete.Limited limit(long limit) {
            return new TypeQLDelete.Limited(this, limit);
        }
    }

    public static class Sorted extends TypeQLDelete implements TypeQLQuery.Sorted<TypeQLDelete.Offset, TypeQLDelete.Limited> {

        public Sorted(TypeQLDelete delete, Modifiers.Sorting sorting) {
            super(delete.match, delete.statements, new Modifiers(sorting, delete.modifiers.offset, delete.modifiers.limit));
            TypeQLQuery.validateSorting(match, sorting);
        }

        @Override
        public TypeQLDelete.Offset offset(long offset) {
            return new TypeQLDelete.Offset(this, offset);
        }

        @Override
        public TypeQLDelete.Limited limit(long limit) {
            return new TypeQLDelete.Limited(this, limit);
        }
    }

    public static class Offset extends TypeQLDelete implements TypeQLQuery.Offset<TypeQLDelete.Limited> {

        public Offset(TypeQLDelete delete, long offset) {
            super(delete.match, delete.statements, new Modifiers(delete.modifiers.sorting, offset, delete.modifiers.limit));
        }

        @Override
        public TypeQLDelete.Limited limit(long limit) {
            return new TypeQLDelete.Limited(this, limit);
        }
    }

    public static class Limited extends TypeQLDelete implements TypeQLQuery.Limited {

        public Limited(TypeQLDelete delete, long limit) {
            super(delete.match, delete.statements, new Modifiers(delete.modifiers.sorting, delete.modifiers.offset, limit));
        }
    }
}

File: rust/query/typeql_delete.rs
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use std::{collections::HashSet, fmt};

use crate::{
    common::{
        error::{collect_err, TypeQLError},
        token,
        validatable::Validatable,
        Result,
    },
    pattern::{ThingStatement, VariablesRetrieved},
    query::{modifier::Modifiers, writable::validate_non_empty, MatchClause, Sorting, TypeQLUpdate, Writable},
    write_joined,
};
use crate::variable::variable::VariableRef;

#[derive(Debug, Eq, PartialEq)]
pub struct TypeQLDelete {
    pub match_clause: MatchClause,
    pub statements: Vec<ThingStatement>,
    pub modifiers: Modifiers,
}

impl TypeQLDelete {
    pub fn insert(self, writable: impl Writable) -> TypeQLUpdate {
        TypeQLUpdate { query_delete: self, insert_statements: writable.statements(), modifiers: Default::default() }
    }

    pub fn sort(self, sorting: impl Into<Sorting>) -> Self {
        TypeQLDelete { modifiers: self.modifiers.sort(sorting), ..self }
    }

    pub fn limit(self, limit: usize) -> Self {
        TypeQLDelete { modifiers: self.modifiers.limit(limit), ..self }
    }

    pub fn offset(self, offset: usize) -> Self {
        TypeQLDelete { modifiers: self.modifiers.offset(offset), ..self }
    }
}

impl Validatable for TypeQLDelete {
    fn validate(&self) -> Result {
        let match_variables = self.match_clause.retrieved_variables().collect();
        collect_err(
            ([
                self.match_clause.validate(),
                validate_delete_in_scope(&match_variables, &self.statements),
                validate_non_empty(&self.statements),
                self.modifiers.sorting.as_ref().map(|s| s.validate(&match_variables)).unwrap_or(Ok(())),
            ]
            .into_iter())
            .chain(self.statements.iter().map(Validatable::validate)),
        )
    }
}

fn validate_delete_in_scope(scope_variables: &HashSet<VariableRef<'_>>, statements: &[ThingStatement]) -> Result {
    collect_err(statements.iter().flat_map(|v| v.variables()).filter(|r| r.is_name()).map(|r| -> Result {
        if scope_variables.contains(&r) {
            Ok(())
        } else {
            Err(TypeQLError::DeleteVarNotBound { variable: r.to_owned() })?
        }
    }))
}

impl fmt::Display for TypeQLDelete {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}", self.match_clause)?;
        writeln!(f, "{}", token::Clause::Delete)?;
        write_joined!(f, ";\n", self.statements)?;
        f.write_str(";")?;
        if !self.modifiers.is_empty() {
            write!(f, "\n{}", self.modifiers)
        } else {
            Ok(())
        }
    }
}
